src/filesystem/impls/lib/handlers.js

```diff
@@ -0,0 +1,76 @@
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, regexp: true, indent: 4, maxerr: 50 */
+/*global define */
+define(function (require, exports, module) {
+    "use strict";
+
+    var Content = require("filesystem/impls/lib/content");
+    var Log = require("filesystem/impls/lib/log");
+    var Rewriter = require("filesystem/impls/lib/rewriter");
+    var Filer = require("thirdparty/filer/dist/filer");
+    var Path  = Filer.Path;
+
+    function handle404(url, callback) {
+        var html = '<!DOCTYPE html>' +
+                   '<html><head>' +
+                   '<title>404 Not Found</title>' +
+                   '</head><body>' +
+                   '<h1>Not Found</h1>' +
+                   '<p>The requested URL ' + url + ' was not found on this server.</p>' +
+                   '<hr>' +
+                   '<address>nohost/0.0.1 (Web) Server</address>' +
+                   '</body></html>';
+        callback(Content.toURL(html, 'text/html'));
+    }
+
+    function handleHTML(fs, path, content, callback) {
+        Rewriter.rewriteHTML(fs, path, content, function(err, html) {
+            if(err) {
+                Log.error('unable to read html for `' + path + '`');
+                // TODO: best way to deal with error here? 500?
+                return handle404(path, callback);
+            }
+
+            callback(Content.toURL(html));
+        });
+    }
+
+    function handleCSS(fs, path, content, callback) {
+        Rewriter.rewriteCSS(fs, path, content, function(err, css) {
+            if(err) {
+                Log.error('unable to read css for `' + path + '`');
+                // TODO: best way to deal with error here? 500?
+                return handle404(path, callback);
+            }
+
+            callback(Content.toURL(css));
+        });
+    }
+
+    /**
+     * Send the raw file, making it somewhat more readable
+     */
+    function handleFile(fs, path, callback) {
+        var ext = Path.extname(path);
+        var encoding = Content.isUTF8Encoded(ext) && 'utf8';
+
+        fs.readFile(path, encoding, function(err, data) {
+            if(err) {
+                Log.error('unable to read `' + path + '` in handler');
+                return handle404(path, callback);
+            }
+
+            if(Content.isHTML(ext)) {
+                handleHTML(fs, path, data, callback);
+            } else if(Content.isCSS(ext)) {
+                handleCSS(fs, path, data, callback);
+            } else {
+                callback(Content.toURL(data));
```

Comment:
Why are you passing this back as the first argument to the callback?  Shouldn't it be the second?